src/main/java/org/opentripplanner/graph_builder/linking/SimpleStreetSplitter.java

```diff
@@ -227,124 +245,86 @@ public boolean link(Vertex vertex, TraverseMode traverseMode, RoutingRequest opt
-
-        // Make a map of distances to all edges.
-        final TObjectDoubleMap<Edge> distances = new TObjectDoubleHashMap<>();
-        for (StreetEdge e : candidateEdges) {
-            distances.put(e, distance(vertex, e, xscale));
-        }
-
-        // Sort the list.
-        Collections.sort(candidateEdges, (o1, o2) -> {
-            double diff = distances.get(o1) - distances.get(o2);
-            // A Comparator must return an integer but our distances are doubles.
-            if (diff < 0)
-                return -1;
-            if (diff > 0)
-                return 1;
-            return 0;
-        });
-
-        // find the closest candidate edges
-        if (candidateEdges.isEmpty() || distances.get(candidateEdges.get(0)) > radiusDeg) {
-            // We only link to stops if we are searching for origin/destination and for that we need transitStopIndex.
-            if (destructiveSplitting || transitStopIndex == null) {
-                return false;
-            }
-            LOG.debug("No street edge was found for {}", vertex);
-            // We search for closest stops (since this is only used in origin/destination linking if no edges were found)
-            // in the same way the closest edges are found.
-            List<TransitStopVertex> candidateStops = new ArrayList<>();
-            transitStopIndex.query(env).forEach(candidateStop -> candidateStops.add((TransitStopVertex) candidateStop));
-
-            final TObjectDoubleMap<Vertex> stopDistances = new TObjectDoubleHashMap<>();
-
-            for (TransitStopVertex t : candidateStops) {
-                stopDistances.put(t, distance(vertex, t, xscale));
-            }
-
-            Collections.sort(candidateStops, (o1, o2) -> {
-                    double diff = stopDistances.get(o1) - stopDistances.get(o2);
-                    if (diff < 0) {
-                        return -1;
-                    }
-                    if (diff > 0) {
-                        return 1;
+            traverseModeSet.setWalk(true);
+        }
+        // Scope block to avoid confusing edge-related local variables with stop-related variables below.
+        {
+            // Perform several transformations at once on the edges returned by the index.
+            // Only consider street edges traversable by the given mode and still present in the graph.
+            // Calculate a distance to each of those edges, and keep only the ones within the search radius.
+            List<DistanceTo<StreetEdge>> candidateEdges = idx.query(env).stream()
+                    .filter(StreetEdge.class::isInstance)
+                    .map(StreetEdge.class::cast)
+                    .filter(e -> e.canTraverse(traverseModeSet) && e.getToVertex().getIncoming().contains(e))
+                    .map(e -> new DistanceTo<>(e, distance(vertex, e, xscale)))
+                    .filter(ead -> ead.distanceDegreesLat < radiusDeg)
+                    .collect(Collectors.toList());
```

Comment:
What is the purpose of `e.getToVertex().getIncoming().contains(e)` ? I think we should add a "business method" to edge witch make this easier to understand:
```
boolean Edge.isConnectedToGraph() {
    return getToVertex().getIncoming().contains(this);
}
```


It might not be needed, if so we can delete it.

Reply:
When the SimpleStreetSplitter is in destructiveSplitting mode (during graph construction rather than handling routing requests), it removes edges that have been split and can then re-split the resulting segments, so parts of them are also removed. New edge fragments are added to the spatial index; the edges that were split are removed (disconnected) from the graph but not removed from the spatial index, so for the rest of the linking and splitting operation we must check whether any edge coming out of the index has not been "soft deleted".

I think all of this is compensating for the fact that STRTrees are optimized at construction and read-only, so you can't remove things from them once they're built. That's no longer relevant since we've been using our own hash grid spatial index instead of the STRTree for a while. I think we can turn this check into an assertion, then change the method that removes the edges so it also removes them from the index.